4.preprocess_features/scripts/1.single_cell_qc.py

```diff
@@ -0,0 +1,513 @@
+cleaned_dir.mkdir(exist_ok=True)
+
+# Directory to save qc result CSV files
+qc_results_dir = pathlib.Path("./qc_results")
+qc_results_dir.mkdir(exist_ok=True)
+
+# Directory to save qc figures
+qc_fig_dir = pathlib.Path("./qc_figures")
+qc_fig_dir.mkdir(exist_ok=True)
+
+# Create an empty dictionary to store data frames for each plate
+all_qc_data_frames = {}
+
+# metadata columns to include in output data frame
+metadata_columns = [
+    "Image_Metadata_Plate",
+    "Image_Metadata_Well",
+    "Image_Metadata_Site",
+    "Nuclei_Location_Center_X",
+    "Nuclei_Location_Center_Y",
+    "Image_FileName_DAPI"
+]
+
+# Get a list of the plates to be used for single-cell QC
+plates = [plate.name.split("_")[0] for plate in data_dir.iterdir() if plate.is_dir()]
+
+# Print the list of plates
+print("These are plates identified:")
+for plate in plates:
+    print(plate)
+
+
+# ## Concat all plates together to assess quality control
+
+# In[3]:
+
+
+# Create an empty list to store data frames for each plate
+all_plate_dfs = []
+
+# Iterate through each plate and create the specified data frame
+for plate in plates:
+    # Only process the files that are in the plate names list
+    plate_path = pathlib.Path(f"{data_dir}/{plate}/per_nuclei.parquet")
+
+    # Read the parquet file into a DataFrame
+    qc_df = pd.read_parquet(plate_path)
+
+    # Append the data frame to the list
+    all_plate_dfs.append(qc_df)
+
+# Concatenate data frames for each plate
+concat_df = pd.concat(all_plate_dfs)
+
+# Reset the index so that the index doesn't reset per plate and cause issues downstream
+concat_df.reset_index(drop=True, inplace=True)
+
+print(concat_df.shape)
+concat_df.head()
+
+
+# ## Evaluate the distributions of the features across plates to determine best features to use for QC
+
+# In[4]:
+
+
+# Set the style
+sns.set_style("whitegrid")
+
+# Create subplots with 2 rows and 2 columns
+fig, axes = plt.subplots(1, 3, figsize=(15, 10))
+
+# Flatten the axes array for easy indexing
+axes = axes.flatten()
+
+# Create kdeplot for Nuclei_AreaShape_Area
+sns.kdeplot(
+    data=concat_df,
+    x="Nuclei_AreaShape_Area",
+    hue="Image_Metadata_Plate",
+    palette="viridis",
+    fill=True,
+    common_norm=False,
+    ax=axes[0],  # Set the first subplot
+)
+axes[0].set_title("Area")
+
+# Create kdeplot for FormFactor
+sns.kdeplot(
+    data=concat_df,
+    x="Nuclei_AreaShape_FormFactor",
+    hue="Image_Metadata_Plate",
+    palette="viridis",
+    fill=True,
+    common_norm=False,
+    ax=axes[1],
+    legend=False,
+)
+axes[1].set_title("Form Factor")
+
+# Create kdeplot for Eccentricity
+sns.kdeplot(
+    data=concat_df,
+    x="Nuclei_AreaShape_Eccentricity",
+    hue="Image_Metadata_Plate",
+    palette="viridis",
+    fill=True,
+    common_norm=False,
+    ax=axes[2],
+    legend=False,
+)
+axes[2].set_title("Eccentricity")
+
+plt.suptitle(
+    "Density plots for all plates confirm distributions are similar for each quality control feature",
```

Comment:
Yeah I made it to be a caption since this notebook and data should be reproducible so that the statement stands. I will make it less presumptive though since we can make the statement as a caption after.